File: Test/TestCases/node-addon-api/object/subscript_operator.cs
using NodeApi;

namespace NodeApiTest;

public partial class TestObject
{
    private static JSValue SubscriptGetWithLatin1StyleString(JSCallbackArgs args)
    {
        JSValue obj = args[0];
        byte[] key = args[1].GetValueStringLatin1();
        return obj[JSValue.CreateStringLatin1(key)];
    }

    private static JSValue SubscriptGetWithUtf8StyleString(JSCallbackArgs args)
    {
        JSValue obj = args[0];
        byte[] key = args[1].GetValueStringUtf8();
        return obj[key];
    }

    private static JSValue SubscriptGetWithCSharpStyleString(JSCallbackArgs args)
    {
        JSValue obj = args[0];
        string key = (string)args[1];
        return obj[key];
    }

    private static JSValue SubscriptGetAtIndex(JSCallbackArgs args)
    {
        JSValue obj = args[0];
        uint index = (uint)args[1];
        return obj[index];
    }

    private static JSValue SubscriptSetWithLatin1StyleString(JSCallbackArgs args)
    {
        JSValue obj = args[0];
        byte[] key = args[1].GetValueStringLatin1();
        JSValue value = args[2];
        obj[JSValue.CreateStringLatin1(key)] = value;
        return JSValue.Undefined;
    }

    private static JSValue SubscriptSetWithUtf8StyleString(JSCallbackArgs args)
    {
        JSValue obj = args[0];
        byte[] key = args[1].GetValueStringUtf8();
        JSValue value = args[2];
        obj[key] = value;
        return JSValue.Undefined;
    }

    private static JSValue SubscriptSetWithCSharpStyleString(JSCallbackArgs args)
    {
        JSValue obj = args[0];
        string key = (string)args[1];
        JSValue value = args[2];
        obj[key] = value;
        return JSValue.Undefined;
    }

    private static JSValue SubscriptSetAtIndex(JSCallbackArgs args)
    {
        JSValue obj = args[0];
        uint index = (uint)args[1];
        JSValue value = args[2];
        obj[index] = value;
        return JSValue.Undefined;
    }
}

File: test/TestCases/node-addon-api/object/subscript_operator.js
'use strict';

const assert = require('assert');

module.exports = require('../../common').runTest(test);

function test(binding) {
  function testProperty(obj, key, value, nativeGetProperty, nativeSetProperty) {
    nativeSetProperty(obj, key, value);
    assert.strictEqual(nativeGetProperty(obj, key), value);
  }

  testProperty({}, 'key', 'value', binding.object.subscriptGetWithLatin1StyleString, binding.object.subscriptSetWithLatin1StyleString);
  testProperty({}, 'key', 'value', binding.object.subscriptGetWithUtf8StyleString, binding.object.subscriptSetWithUtf8StyleString);
  testProperty({ key: 'override me' }, 'key', 'value', binding.object.subscriptGetWithCSharpStyleString, binding.object.subscriptSetWithCSharpStyleString);
  testProperty({}, 0, 'value', binding.object.subscriptGetAtIndex, binding.object.subscriptSetAtIndex);
  testProperty({ key: 'override me' }, 0, 'value', binding.object.subscriptGetAtIndex, binding.object.subscriptSetAtIndex);
}
